### PR TITLE
Docgen

### PR DESCRIPTION
**Pull request by Conni2461**
Take 2 :|

### CODE REVIEW

File: scripts/gendocs.lua [hunk 1]
```diff
@@ -0,0 +1,22 @@
+local docgen = require'docgen'
+
+local function gen()
+  local input_files = {
+    './lua/sql.lua',
+    './lua/sql/table.lua',
+    './lua/sql/stmt.lua',
```
Comment (Conni2461) [inline scripts/gendocs.lua:7]: @tami5 for which files do you want to generate docs?

Also i think most files don't have a brief
```lua
---@brief [[
--- Desc
---@brief ]]
```

Comment (Conni2461) [reply scripts/gendocs.lua:7]: I am not sure. I haven't touched that table since i started this PR a "year" ago.

I just rebased :rofl:

File: doc/sql.txt [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+                                                                       *sql.lua*
+
+SQLite/LuaJIT binding and highly opinionated wrapper for storing, retrieving,
+caching, persisting, querying, and connecting to SQLite databases.
+To find out more:
+https://github.com/tami5/sql.nvim
+  :h sql
+  :h sql.table
+
+sql.new({uri})                                                     *sql.new()*
```
Comment (tami5) [inline doc/sql.txt:11]: idk it feels wrong to call it sql, db might be better but that something for another pr

File: doc/sql.txt [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+        {params} (table)  params to be bind to {statement}, it can be a list
+                          or dict
+
+    Return: ~
+        boolean or table
+
+    Usage: ~
+        db:eval("drop table if exists todos")
+        db:eval("select * from todos where id = ?", 1)
+        db:eval("insert into t(a, b) values(:a, :b)", {a = "1", b = 2021})
```
Comment (tami5) [inline doc/sql.txt:102]: wrap with "`"

File: doc/sql.txt [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+    Parameters: ~
+        {name} (string)  the table name.
+
+    Return: ~
+        boolean
+
+
+sql:schema({tbl}, {info})                                       *sql:schema()*
+    get sql table {name} schema, if table doesn't exist then return empty
+    table.
```
Comment (tami5) [inline doc/sql.txt:139]:
```suggestion
    Get {name} table schema, if table doesn't exist then return empty
    table.
```

File: doc/sql.txt [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+
+    Return: ~
+        boolean
+
+    Usage: ~
+        `db:create("todos", {id = {"int", "primary", "key"}, title = "text"})`
+
+
+sql:drop({name})                                                  *sql:drop()*
+    Create a new sqlite db table with {name} based on {schema}. if
```
Comment (tami5) [inline doc/sql.txt:168]: wrong doc. not docgen part I believe

File: doc/sql.txt [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+        {name} (string)  table name
+
+    Return: ~
+        boolean
+
+    Usage: ~
+        `db:drop("todos")`
+
+
+sql:insert()                                                    *sql:insert()*
```
Comment (tami5) [inline doc/sql.txt:183]: missing args

File: doc/sql.txt [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+    Insert to lua table into sqlite database table.
+
+
+    Return: ~
+        boolean|integer: true incase the table was inserted successfully, and
+        the last inserted row id.
+
+    Usage: ~
+        db:insert("todos", { title = "new todo" })
```
Comment (tami5) [inline doc/sql.txt:192]: wrap with "`"

File: doc/sql.txt [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+
+
+sql:update({tbl}, {specs})                                      *sql:update()*
+    Update table row with where closure and list of values
+
+
+    Parameters: ~
```
Comment (tami5) [inline doc/sql.txt:199]: tbl is a ambiguous, must be changed to tbl_name instead, and same goes for other functions

File: doc/sql.txt [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+        {tbl}   (string)  the name of the db table.
+        {specs} (table)   a {spec} or a list of {specs} with where and values
+                          key.
+
+    Return: ~
+        boolean: true incase the table was updated successfully.
+
+    Usage: ~
+        `db:update("todos", { where = { id = "1" }, values = { action = "DONE"
+        }})`
```
Comment (tami5) [inline doc/sql.txt:209]: should we change it to be multiline?

File: doc/sql.txt [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+        {specs} (table)   a {spec} or a list of {specs} with where and values
+                          key.
+
+    Return: ~
+        boolean: true if operation is successfully, false otherwise.
+
+    Usage: ~
+        db:delete("todos")
+        db:delete("todos", { where = { id = 1 })
+        db:delete("todos", { where = { id = {1,2,3} })
```
Comment (tami5) [inline doc/sql.txt:228]: wrap with "`"

File: doc/sql.txt [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+
+
+sql:select()                                                    *sql:select()*
```
Comment (tami5) [inline doc/sql.txt:231]: missing args

File: doc/sql.txt [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+    Query from a table with where and join options
+
+
+    Return: ~
+        lua list of matching rows
+
+    Usage: ~
+        db:get("todos") -- everything
+        db:get("todos", { where = { id = 1 })
+        db:get("todos", { limit = 5 })
```
Comment (tami5) [inline doc/sql.txt:241]: should be db:select

File: doc/sql.txt [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+
+
+    Parameters: ~
+        {where} (table)  where key values
+
+    Return: ~
+        nil or row
+
+    Usage: ~
+        ---@usage: tbl:where{id = 1}
```
Comment (tami5) [inline doc/sql.txt:330]: maybe bug

File: doc/sql.txt [hunk 14]
```diff
@@ -0,0 +1,436 @@
...
+
+    See: ~
+        |sql:select()|
+
+
+tbl:each({query}, {func})                                         *tbl:each()*
+    Iterate over {self.name} rows and execute {func}.
```
Comment (tami5) [inline doc/sql.txt:337]:
```suggestion
    Iterate over table rows and execute {func}.
```

File: doc/sql.txt [hunk 15]
```diff
@@ -0,0 +1,436 @@
...
+    See: ~
+        |sql:insert()|
+
+
+tbl:remove({specs})                                             *tbl:remove()*
+    Same functionalities as |sql:delete()|
+
+
+    Parameters: ~
+        {specs} (table)  specs.where
```
Comment (tami5) [inline doc/sql.txt:395]: should rename to where?

File: doc/sql.txt [hunk 16]
```diff
@@ -0,0 +1,436 @@
...
+
+    Return: ~
+        boolean
+
+    See: ~
+        |sql:update()|
+
+
+tbl:replace({rows})                                            *tbl:replace()*
+    Same functionalities as |tbl:add()|, but replaces {self.name} content with
```
Comment (tami5) [inline doc/sql.txt:420]:
```suggestion
    Same functionalities as |tbl:add()|, but replaces table content with
```

File: lua/sql/stmt.lua [hunk 17]
```diff
@@ -5,33 +5,26 @@ local Stmt = {}
...
----@see Stmt:__parse()
----@usage local stmt = Stmt:parse(db, "insert into todos (title,desc) values(:title, :desc)")
-function Stmt:parse(conn, stmt)
-  assert(sqlite.type_of(conn) == sqlite.type_of_db_ptr, "Invalid connection passed to stmt:parse")
-  assert(type(stmt) == "string", "Invalid second argument passed to stmt:parse")
+---@tag Stmt.lua
+
+---@class sqlite3 @sqlite3 db object
+---@class sqlite3_blob @sqlite3 blob object
+---@class sqlite3_flag @sqlite3 error flag
```
Comment (tami5) [inline lua/sql/stmt.lua:14]: should be moved to defs.lua

File: lua/sql/stmt.lua [hunk 18]
```diff
@@ -5,33 +5,26 @@ local Stmt = {}
...
+
+---Create new object for {conn} to deal with sqlite {Stmt}
+---@class Stmt @oject to deal with sqlite statements
```
Comment (tami5) [inline lua/sql/stmt.lua:17]: oject?

File: lua/sql/stmt.lua [hunk 19]
```diff
@@ -139,11 +123,11 @@ local sqlite_datatypes = {
   [5] = "null", -- bind_null
 }
 
---- Key/Column lua datatype at {idx}
+---Key/Column lua datatype at {idx}
 ---@param idx number: (0-index)
 ---@return string: key/column type at {idx}
 ---@TODO should accept 1-index
-function Stmt:type(idx)
+function Stmt:convert_type(idx)
```
Comment (tami5) [inline lua/sql/stmt.lua:130]: 👍🏻👍🏻

**Review by tami5 — CHANGES_REQUESTED** [review]
hmmm I'm missing a lot of examples maybe it's something for another pr?

File: doc/sql.txt (hunk 13 above)
Comment (Conni2461) [reply doc/sql.txt:330]: maybe maybe not. it just doesn't like `:` i remove it ;)

File: doc/sql.txt (hunk 12 above)
Comment (Conni2461) [reply doc/sql.txt:241]: i only added the ci job :sob: now i am fixing your documentation :rofl:

File: doc/sql.txt (hunk 11 above)
Comment (Conni2461) [reply doc/sql.txt:231]: no docs no arguments generated :laughing:

Comment (Conni2461) [reply doc/sql.txt:231]: a question about this. Shouldn't this work the same as `delete`? but there you also allow a list of specs. Not here

File: doc/sql.txt (hunk 9 above)
Comment (Conni2461) [reply doc/sql.txt:209]: i haven't gotten around to implement multiline emmy annotation. Its actually kinda hard to do :grimacing:

File: doc/sql.txt (hunk 6 above)
Comment (Conni2461) [reply doc/sql.txt:183]: nothing docgen can do if you annotate your parameter with `@params`. Only `@param` is supported :rofl:

File: doc/sql.txt (hunk 5 above)
Comment (Conni2461) [reply doc/sql.txt:168]: ???

File: doc/sql.txt (hunk 2 above)
Comment (Conni2461) [reply doc/sql.txt:11]: You know that we then have to rename this: https://github.com/tami5/sql.nvim/blob/3a51b7b9780d76f8f2feef7e2034fb6fe4cf3e60/lua/sql.lua#L20

Comment (tami5) [reply doc/sql.txt:11]: yep yep makes sense to do so will do later

File: doc/sql.txt (hunk 6 above)
Comment (tami5) [reply doc/sql.txt]: 😂😂😂😂

**Review by tami5 — APPROVED** [review]
LGTM, finally docss
